js/src/data/reducer.test.js

```diff
@@ -0,0 +1,641 @@
+/**
+ * External dependencies
+ */
+import { cloneDeep, set, get, isPlainObject } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import reducer from './reducer';
+import TYPES from './action-types';
+
+// Copied from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze
+function deepFreeze( object ) {
+	// Retrieve the property names defined on object
+	const propNames = Object.getOwnPropertyNames( object );
+
+	// Freeze properties before freezing self
+	for ( const name of propNames ) {
+		const value = object[ name ];
+		if ( value && typeof value === 'object' ) {
+			deepFreeze( value );
+		}
+	}
+
+	return Object.freeze( object );
+}
+
+function attachRef( object, ignore, checkingGroups = [], paths = [] ) {
+	const refKey = '__refForCheck';
+
+	for ( const [ name, value ] of Object.entries( object ) ) {
+		if ( isPlainObject( value ) ) {
+			const checkPaths = [ ...paths, name ];
+			attachRef( value, ignore, checkingGroups, checkPaths );
+
+			const currentPath = checkPaths.join( '.' );
+			if ( ! ignore.includes( currentPath ) ) {
+				const ref = {
+					whyFail: `If there's no mutation at \`state.${ currentPath }\`, it should be kept the same reference.`,
+				};
+				checkingGroups.push( {
+					ref,
+					refPath: `${ currentPath }.${ refKey }`,
+				} );
+				value[ refKey ] = ref;
+			}
+		}
+	}
+	return checkingGroups;
+}
+
+describe( 'reducer', () => {
+	let defaultState;
+	let prepareState;
+
+	beforeEach( () => {
+		defaultState = deepFreeze( {
+			mc: {
+				target_audience: null,
+				countries: null,
+				shipping: {
+					rates: [],
+					times: [],
+				},
+				settings: null,
+				accounts: {
+					jetpack: null,
+					google: null,
+					mc: null,
+					ads: null,
+					existing_mc: null,
+					existing_ads: null,
+					ads_billing_status: null,
+					google_access: null,
+				},
+				contact: null,
+			},
+			ads_campaigns: null,
+			mc_setup: null,
+			mc_product_statistics: null,
+			mc_issues: null,
+			mc_product_feed: null,
+			report: {},
```

Comment:
Ideally this should be imported from `reducer.js`  instead of duplicating the code. Otherwise a part of the verbosity, each time we modify something in reducer default state we need to do it here as well.

Reply:
> Ideally this should be imported from `reducer.js` instead of duplicating the code. Otherwise a part of the verbosity, each time we modify something in reducer default state we need to do it here as well.

and quotes from https://github.com/woocommerce/google-listings-and-ads/pull/1150#discussion_r768017840:

> Besides that, It would be good to have the defaultState imported from DEFAULT_STATE constant instead of hardcoding it in the beforeEach statement.

Sorry, I don't think the `defaultState` should be imported from `DEFAULT_STATE`. Testing the default values also allows us to check the shape of state tree and each initial value for unintended changes, and there are many logics that rely on state's initial values.

If we import the same source (`DEFAULT_STATE`) from reducer, then all reducer tests are still passed even I change the almost whole state tree by deleting nodes, renaming node names, and changing to other types like this:

```js
export const DEFAULT_STATE = {
	mc: {
		// deleted some nodes
		userSettings: null,
		accounts: [], // changed from object to array
		contact: NaN, // changed from null to NaN
	},
	// rename all remaining nodes
	campaigns: null,
	setups: null,
	product_statistics: null,
	issues: null,
	product_feed: null,
	reports: {},
};
```

and its diff:

```diff
 const DEFAULT_STATE = {
        mc: {
-               target_audience: null,
-               countries: null,
-               shipping: {
-                       rates: [],
-                       times: [],
-               },
-               settings: null,
-               accounts: {
-                       jetpack: null,
-                       google: null,
-                       mc: null,
-                       ads: null,
-                       existing_mc: null,
-                       existing_ads: null,
-                       ads_billing_status: null,
-                       google_access: null,
-               },
-               contact: null,
+               userSettings: null,
+               accounts: [],
+               contact: NaN,
        },
-       ads_campaigns: null,
-       mc_setup: null,
-       mc_product_statistics: null,
-       mc_issues: null,
-       mc_product_feed: null,
-       report: {},
+       campaigns: null,
+       setups: null,
+       product_statistics: null,
+       issues: null,
+       product_feed: null,
+       reports: {},
 };
```

Reply:
I'm happy to move the hardcoded `defaultState` to another file, that is under a jest convention folder such as `__fixtures__`, if you think it has bad smell of readability.

Reply:
Makes sense. Didn't thought about you wanted to protect default_state changes. I checked removing some properties in the default_state (app) and run the test and it failed in the `it( 'should return default state by default' ` assertion. So, all good